### PR TITLE
 shing basics 11-7 beat that

### PR DESCRIPTION
**Pull request by shingoh24**
Please fill out the survey before submitting the pull request. Thanks!

🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀

How many hours did you spend on this assignment?
3 hrs or more... 

Please fill in one error and/or error message you received while working on this assignment.
NaN when concatenating combined dice roll 
 
What part of the assignment did you spend the most time on?
debugging the concatenating error but still failed 
Comfort Level (1-5): 

Completeness Level (1-5): 

What did you think of this deliverable?

Is there anything in this code that you feel pleased about?

### CODE REVIEW

File: script.js [hunk 1]
```diff
@@ -1,4 +1,150 @@
...
+      "<br> Please enter ❶ or ❷ as the first numeral index for your dice rolls to get the highest combined number and hit the submit button.";
+  } else if (currentGameMode == CHOOSE_COMBINATION_PLAYER) {
+    // Validate the input. If first numeral index is neither 1 nor 2, tell the user.
+    var firstNumeralIndex = Number(input);
+    if (firstNumeralIndex !== 1 && firstNumeralIndex !== 2) {
+      return "Please choose ❶ or ❷ as the first numeral index for your dice rolls";
+    }
+    // Get combined number for current player
+    var playerCombined = getPlayerNumber(firstNumeralIndex);
+    console.log(playerCombined);
```
Comment (leechuanxin) [inline script.js]: I saw that you `console.log(playerCombined)` here, and that's where you first found out about the  `NaN` error.

Good start on learning debugging! Only small thing I will note, I will try to console.log it together with a string, just like this:
```
console.log("playerCombined: " + playerCombined);
```
This will allow me to associate the `NaN` value that is printed with the name of the variable storing this erroneous value, when I view the console.logs from the browser console.

File: script.js [hunk 2]
```diff
@@ -1,4 +1,150 @@
...
+      " and " +
+      playerDiceRollTwo +
+      "<br> Please enter ❶ or ❷ as the first numeral index for your dice rolls to get the highest combined number and hit the submit button.";
+  } else if (currentGameMode == CHOOSE_COMBINATION_PLAYER) {
+    // Validate the input. If first numeral index is neither 1 nor 2, tell the user.
+    var firstNumeralIndex = Number(input);
+    if (firstNumeralIndex !== 1 && firstNumeralIndex !== 2) {
+      return "Please choose ❶ or ❷ as the first numeral index for your dice rolls";
+    }
+    // Get combined number for current player
```
Comment (leechuanxin) [inline script.js:103]: I know from the previous `console.log` that you have written that the `playerCombined` variable in [line 103](https://github.com/rocketacademy/basics-beat-that/blob/423b4b54b25fc979d6590b7cc737c2fcf9b09b53/script.js#L103) is problematic.

`playerCombined` stores the `return` value from the execution of `getPlayerNumber(firstNumeralIndex)`. Since I am getting `NaN`, either the function itself is problematic, or the parameter to the function `firstNumeralIndex` is problematic.

I will start by testing the parameter `firstNumeralIndex`. Somewhere after [line 102](https://github.com/rocketacademy/basics-beat-that/blob/423b4b54b25fc979d6590b7cc737c2fcf9b09b53/script.js#L102), but before [line 103 ](https://github.com/rocketacademy/basics-beat-that/blob/423b4b54b25fc979d6590b7cc737c2fcf9b09b53/script.js#L103), I will console.log `firstNumeralIndex`, like so:

```
console.log("firstNumeralIndex: " + firstNumeralIndex);
```

Save and refresh your browser, and test it out.

File: script.js [hunk 3]
```diff
@@ -1,4 +1,150 @@
...
+
+  var playerNum;
+  // If the chosen first numeral index is 1, create player number starting with 1st dice
+  // Otherwise, create player number starting with 2nd dice
+  if (firstNumeralIndex === 1) {
+    playerNum = concatenateTwoNumbers(diceArray[0], diceArray[1]);
+  } else {
+    playerNum = concatenateTwoNumbers(diceArray[1], diceArray[0]);
+  }
+
```
Comment (leechuanxin) [inline script.js:56]: I have tested `firstNumeralIndex` earlier, and it seems like it is giving me valid numbers of either 1 or 2. So, let's assume it is okay for now. That means something must be wrong with the `getPlayerNumber` function. Now, I am looking straight into lines of code where the function is created (function definition). My eyes are staring between [lines 38 and 64](https://github.com/rocketacademy/basics-beat-that/blob/423b4b54b25fc979d6590b7cc737c2fcf9b09b53/script.js#L38-L64) now.

I saw that the function `return`s `playerNum` at the end, and you are assigning `playerNum` using `if / else` blocks between [lines 49 and 53](https://github.com/rocketacademy/basics-beat-that/blob/423b4b54b25fc979d6590b7cc737c2fcf9b09b53/script.js#L49-L53). 

You can also drop a `console.log` right around the blank line at [line 54](https://github.com/rocketacademy/basics-beat-that/blob/423b4b54b25fc979d6590b7cc737c2fcf9b09b53/script.js#L54) here, just to double-confirm it is `NaN` like `playerCombined` in [line 104](https://github.com/rocketacademy/basics-beat-that/blob/423b4b54b25fc979d6590b7cc737c2fcf9b09b53/script.js#L104).

```
console.log("playerNum: " + playerNum);
```

Save your file, refresh your browser, and test.

File: script.js [hunk 4]
```diff
@@ -1,4 +1,150 @@
...
+    var diceArray = playerOneDice;
+  } else {
+    var diceArray = playerTwoDice;
+  }
+
+  var playerNum;
+  // If the chosen first numeral index is 1, create player number starting with 1st dice
+  // Otherwise, create player number starting with 2nd dice
+  if (firstNumeralIndex === 1) {
+    playerNum = concatenateTwoNumbers(diceArray[0], diceArray[1]);
```
Comment (leechuanxin) [inline script.js:51]: If you have been following till now (debugging requires a lot of patience, doesn't it?), `playerNum` is still going to be `NaN`. Now, I see on [line 46](https://github.com/rocketacademy/basics-beat-that/blob/423b4b54b25fc979d6590b7cc737c2fcf9b09b53/script.js#L46) that `playerNum` is created with no values assigned, so it is probably in the lines that it gets assigned values which are problematic. Now we have narrowed it down to 2 lines: lines [50](https://github.com/rocketacademy/basics-beat-that/blob/423b4b54b25fc979d6590b7cc737c2fcf9b09b53/script.js#L50) and [52](https://github.com/rocketacademy/basics-beat-that/blob/423b4b54b25fc979d6590b7cc737c2fcf9b09b53/script.js#L52).

So we realise from these 2 lines that `playerNum` is going to be re-assigned values from the `return` value from the execution of the `concatenateTwoNumbers` function. In lines [50](https://github.com/rocketacademy/basics-beat-that/blob/423b4b54b25fc979d6590b7cc737c2fcf9b09b53/script.js#L50) and [52](https://github.com/rocketacademy/basics-beat-that/blob/423b4b54b25fc979d6590b7cc737c2fcf9b09b53/script.js#L52), we know we are passing in array elements from `diceArray` as parameters into the `concatenateTwoNumbers` function for both the `(firstNumeralIndex === 1)` and the `else` cases.

The `return` of executing `concatenateTwoNumbers` is problematic, and just like how I tested `getPlayerNumber` earlier: when a `return` value from the execution of a function is problematic, it is usually the function itself being problematic, or the parameters.

For starters, let me test the parameters. Your parameters `diceArray[0]` and `diceArray[1]` are array elements, but before I test the array elements, let me test the entire array first. 

Maybe right around the blank line on [line 45](https://github.com/rocketacademy/basics-beat-that/blob/423b4b54b25fc979d6590b7cc737c2fcf9b09b53/script.js#L45), let's console.log `diceArray`?

```
console.log("diceArray: " + diceArray);
```

Save your file, refresh your browser, and test.

File: script.js [hunk 5]
```diff
@@ -1,4 +1,150 @@
...
+console.log("two numbers");
+
+var getPlayerNumber = function (firstNumeralIndex) {
+  var diceArray;
+  if (currentPlayer === 1) {
+    var diceArray = playerOneDice;
+  } else {
+    var diceArray = playerTwoDice;
+  }
+
```
Comment (leechuanxin) [inline script.js:46]: Aha! If you have done your `console.log` of the `diceArray` on line 45 as I told you and tested your game, you will realise it is always an empty array.

This means there's something wrong with the assignment of `diceArray`. Despite [line 41](https://github.com/rocketacademy/basics-beat-that/blob/423b4b54b25fc979d6590b7cc737c2fcf9b09b53/script.js#L41) assigning it to `playerOneDice` and [line 43](https://github.com/rocketacademy/basics-beat-that/blob/423b4b54b25fc979d6590b7cc737c2fcf9b09b53/script.js#L43) assigning it to `playerTwoDice`, it is still empty.

`playerOneDice` and `playerTwoDice` are likely problematic. Firstly, I will check where these 2 variables are created. I will do a Ctrl+F of either "var playerOneDice" or "var playerTwoDice" to check where they are created. I learnt that they are created around [lines 12 and 13](https://github.com/rocketacademy/basics-beat-that/blob/423b4b54b25fc979d6590b7cc737c2fcf9b09b53/script.js#L12-L13).

They seem to be created perfectly okay. Just 2 variables which initially store empty arrays. Now, I will look to see where these variables change, ie. where these variables are re-assigned. I will do a Ctrl+F of "playerOneDice =" or "playerTwoDice =", without the `var` (since we are talking about re-assignment of values). I found 2 places where they are re-assigned: [line 26](https://github.com/rocketacademy/basics-beat-that/blob/423b4b54b25fc979d6590b7cc737c2fcf9b09b53/script.js#L26), and [line 28](https://github.com/rocketacademy/basics-beat-that/blob/423b4b54b25fc979d6590b7cc737c2fcf9b09b53/script.js#L28).

Now then, I know that [line 26](https://github.com/rocketacademy/basics-beat-that/blob/423b4b54b25fc979d6590b7cc737c2fcf9b09b53/script.js#L26) and [line 28](https://github.com/rocketacademy/basics-beat-that/blob/423b4b54b25fc979d6590b7cc737c2fcf9b09b53/script.js#L28) are within the getDiceRolls function from [lines 23 to 31](https://github.com/rocketacademy/basics-beat-that/blob/423b4b54b25fc979d6590b7cc737c2fcf9b09b53/script.js#L23-L31). Next, I will check where the function is being executed. Let me do a Ctrl+F of "getDiceRolls(" just to check where it is being executed ...

File: script.js [hunk 6]
```diff
@@ -1,4 +1,150 @@
...
 var main = function (input) {
-  var myOutputValue = 'hello world';
+  var myOutputValue = "";
+  if ((input == "") & (currentGameMode == AWAITING_PLAYER_ONE)) {
+    currentGameMode = START_PLAYER_DICE_ROLL;
+    myOutputValue =
+      "Hi Player One, welcome to the 🎲 dice 🎲 roll game! <br> <br> To start, please hit the submit button to roll the dice 🎲!";
+  } else if (currentGameMode == START_PLAYER_DICE_ROLL) {
+    var playerDiceRollOne = diceRollOne();
```
Comment (leechuanxin) [inline script.js]: And I realised "getDiceRolls(" is not found anywhere on this script.js at all!

This means the [lines 26](https://github.com/rocketacademy/basics-beat-that/blob/423b4b54b25fc979d6590b7cc737c2fcf9b09b53/script.js#L26) and [28](https://github.com/rocketacademy/basics-beat-that/blob/423b4b54b25fc979d6590b7cc737c2fcf9b09b53/script.js#L28) we want executed have never been executed, because `getDiceRolls` is never executed!

Now, then I will try to find the place where I think `getDiceRolls` should be executed, which is probably around where the `currentGameMode == START_PLAYER_DICE_ROLL`.

Now that we realise we have not been executing `getDiceRolls` in this `else if` block, let's change your code a little so that we execute it now.

You initialise your dice rolls currently on [lines 85 and 86](https://github.com/rocketacademy/basics-beat-that/blob/423b4b54b25fc979d6590b7cc737c2fcf9b09b53/script.js#L85-L86). You realise that it is the same as what you are doing in [line 24](https://github.com/rocketacademy/basics-beat-that/blob/423b4b54b25fc979d6590b7cc737c2fcf9b09b53/script.js#L24) within the `getDiceRolls` function?

We also realise on [line 30](https://github.com/rocketacademy/basics-beat-that/blob/423b4b54b25fc979d6590b7cc737c2fcf9b09b53/script.js#L30) in `getDiceRolls` that we are returning an array created [here](https://github.com/rocketacademy/basics-beat-that/blob/423b4b54b25fc979d6590b7cc737c2fcf9b09b53/script.js#L24).

Thus, we do not need the variables `playerDiceRollOne` and `playerDiceRollTwo` anymore. Let's combine them into 1 variable called `playerDiceRolls` which stores both rolls as an array that is [returned](https://github.com/rocketacademy/basics-beat-that/blob/423b4b54b25fc979d6590b7cc737c2fcf9b09b53/script.js#L30) from `getDiceRolls`.

We can delete [lines 85 and 86](https://github.com/rocketacademy/basics-beat-that/blob/423b4b54b25fc979d6590b7cc737c2fcf9b09b53/script.js#L85-L86), and replace them with this:

```
var playerDiceRolls = getDiceRolls();
```

File: script.js [hunk 7]
```diff
@@ -1,4 +1,150 @@
...
+    var playerDiceRollTwo = diceRollTwo();
+    currentGameMode = CHOOSE_COMBINATION_PLAYER;
+    myOutputValue =
+      "Hi Player " +
+      currentPlayer +
+      " , you rolled: " +
+      playerDiceRollOne +
+      " and " +
+      playerDiceRollTwo +
+      "<br> Please enter ❶ or ❷ as the first numeral index for your dice rolls to get the highest combined number and hit the submit button.";
```
Comment (leechuanxin) [inline script.js]: And now your copywriting here has to be changed, because the variables `playerDiceRollOne` and `playerDiceRollTwo` don't exist anymore. Instead, we created the variable `playerDiceRolls` from the previous step which stores an array of 2 values.

Now, we want to access the individual values using `playerDiceRolls[0]` and `playerDiceRolls[1]` instead, to get the 2 different dice rolls. Thus, `myOutputValue` should be re-assigned like so:

```
myOutputValue =
      "Hi Player " +
      currentPlayer +
      " , you rolled: " +
      playerDiceRolls[0] +
      " and " +
      playerDiceRolls[1] +
      "<br> Please enter ❶ or ❷ as the first numeral index for your dice rolls to get the highest combined number and hit the submit button.";
```

And ... voila! Go test your game now, it should work :)

**Review by leechuanxin — COMMENTED** [review]
Follow along the `console.log` steps here sequentially to follow my train of thought when looking for the reason for the `NaN`. Yes, sometimes debugging feels like intensive detective work, but I assure you those are the exact steps I have taken to find the problem :). To this day I am still using the same (tedious) strategies for my work. Hope it can enlighten you to follow the same strategies for debugging your subsequent assignments yourself!

File: script.js [hunk 8]
```diff
@@ -1,4 +1,150 @@
...
+  var diceNumber = randomInteger + 1;
+  return diceNumber;
+};
+
+var diceRollTwo = function () {
+  var randomDecimal = Math.random() * 6;
+  var randomInteger = Math.floor(randomDecimal);
+  var diceNumber = randomInteger + 1;
+  return diceNumber;
+};
```
Comment (leechuanxin) [inline script.js:148]: We do not need to have 2 functions `diceRollOne` and `diceRollTwo`. We can just keep it to one `diceRoll` function.

I am assuming you thought you needed to execute 2 different `diceRoll` functions to generate 2 different random numbers. You actually need only execute the **same** `diceRoll` function **twice** to get 2 different random numbers.

Here's how it works - Example 1, an array with 2 separate dice rolls (2 random numbers)
1. Let's assume you either have an array that stores 2 different dice rolls, like so `var diceRollArr = [diceRoll(), diceRoll()];`
2. Firstly, your code will execute the first `diceRoll()` it sees, which is in index 0 of `diceRollArr` or `diceRollArr[0]`.
3. When `diceRoll` is executed, it jumps to [line 141 ](https://github.com/rocketacademy/basics-beat-that/blob/423b4b54b25fc979d6590b7cc737c2fcf9b09b53/script.js#L141) to execute the first line in the function. This line executes the `Math.random` function before multiplying 6. Executing `Math.random` will generate a random number from 0 to 0.9999... every time.
4. Try it for yourself to prove it: go to your browser, open up your browser console / Console tab (Inspect Element). Keep typing in `Math.random()` and hitting Enter. Besides the very very unlikely case where you generate the same decimal twice in a row (we can get randomly lucky), you should get a new random number each time.
4. This means when it is time to execute the `diceRoll()` function the 2nd time (or in index 1 of `diceRollArr` or `diceRollArr[1]`), it will likewise jump back to [line 141 ](https://github.com/rocketacademy/basics-beat-that/blob/423b4b54b25fc979d6590b7cc737c2fcf9b09b53/script.js#L141), which in turn create a completely new random number because it executes `Math.random`.

It works similarly if you decide to create 2 variables to store different dice rolls. Let's imagine Example 2, 2 variables storing 2 separate dice rolls (2 random numbers):

```
var firstDiceRoll = diceRoll();
var secondDiceRoll = diceRoll();
```

1. On the first execution (before assignment of `return` value to `firstDiceRoll`), it goes to [line 141 ](https://github.com/rocketacademy/basics-beat-that/blob/423b4b54b25fc979d6590b7cc737c2fcf9b09b53/script.js#L141) to generate a new random number by executing `Math.random`.
2. On the second execution (before assignment of `return` value to `secondDiceRoll`), it goes back to [line 141 ](https://github.com/rocketacademy/basics-beat-that/blob/423b4b54b25fc979d6590b7cc737c2fcf9b09b53/script.js#L141) to generate yet another new random number by executing `Math.random`.

TL;DR: this means you can delete [lines 140 to 145](https://github.com/rocketacademy/basics-beat-that/blob/423b4b54b25fc979d6590b7cc737c2fcf9b09b53/script.js#L140-L145), and [rename diceRollOne to diceRoll](https://github.com/rocketacademy/basics-beat-that/blob/423b4b54b25fc979d6590b7cc737c2fcf9b09b53/script.js#L133) :D

File: script.js [hunk 9]
```diff
@@ -1,4 +1,150 @@
...
+// Keep track of each player's chosen numbers
+var playerOneCombination;
+var playerTwoCombination;
+
+// Create an array combinedDiceRolls with 2 independent dice roll values
+// Assign combinedDiceRolls to the current player's dice array
+// If currentPlayer is not 1, assume currentPlayer is 2
+// Return combined dice rolls to parent function
+var getDiceRolls = function () {
+  var combinedDiceRolls = [diceRollOne(), diceRollTwo()];
```
Comment (leechuanxin) [inline script.js:24]: Since `diceRollOne` and `diceRollTwo` functions don't exist anymore, and only the newly renamed `diceRoll` function remains, we need to change the functions we are executing:

```
var combinedDiceRolls = [diceRoll(), diceRoll()];
```

File: script.js [hunk 10]
```diff
@@ -1,4 +1,150 @@
...
+  return Number(String(num1) + String(num2));
+};
+console.log("two numbers");
+
+var getPlayerNumber = function (firstNumeralIndex) {
+  var diceArray;
+  if (currentPlayer === 1) {
+    var diceArray = playerOneDice;
+  } else {
+    var diceArray = playerTwoDice;
```
Comment (leechuanxin) [inline script.js]: `diceArray` is already created (see the `var` keyword) on [line 39](https://github.com/rocketacademy/basics-beat-that/blob/423b4b54b25fc979d6590b7cc737c2fcf9b09b53/script.js#L39).

Thus, on lines 41 and 43 here, we do not have to re-create it (ie. no need to use the `var` keyword anymore) to use it. We need only change its value, or re-assign it to another value. 

Thus, replace this:

```
if (currentPlayer === 1) {
  var diceArray = playerOneDice;
} else {
  var diceArray = playerTwoDice;
}
```

With this:

```
if (currentPlayer === 1) {
  diceArray = playerOneDice;
} else {
  diceArray = playerTwoDice;
}
```

**Review by leechuanxin — COMMENTED** [review]
This review is not the one where you requested for debugging help, but I just decided to take a look at the rest of your code to remove certain obvious redundancies.

I want to remark that you have improved quite a lot from your Scissors, Paper, Stone submission. You were struggling a lot with logic back then, and now you are speaking JavaScript more fluently. You are already writing your own functions (ie. not throwing all of your logic within `main` function) like `concatenateTwoNumbers`, `getDiceRolls`, `getPlayerNumber`, to name a few. 

How does it feel knowing that Scissors, Paper, Stone is only 1 week ago? How, then, does it feel to (probably unknowingly) write your first program with a pretty decent 100+ lines? :P

Hope the debugging exercise earlier helped! I won't be evaluating your code too much beyond this for now, because I want you to focus on my current comments. Also, would like you to get a hang of debugging on your own by following my thought processes / detective work earlier.

Of course, if you are still working further to improve your Beat That project over the next few days, just drop me a message again if you get stuck or want another review.
